Content.Server/EnergyDome/EnergyDomeSystem.cs

```diff
@@ -0,0 +1,259 @@
+            _popup.PopupEntity(
+                Loc.GetString("energy-dome-no-cell"),
+                generator,
+                user);
+            return false;
+        }
+
+        if (generator.Comp.NextActivation > _gameTiming.CurTime)
+        {
+            _audio.PlayPvs(generator.Comp.TurnOffSound, generator);
+            _popup.PopupEntity(
+                Loc.GetString("energy-dome-recharging"),
+                generator,
+                user);
+            return false;
+        }
+
+
+
+        if (battery.Charge < generator.Comp.Wattage)
+        {
+            _audio.PlayPvs(generator.Comp.TurnOffSound, generator);
+            _popup.PopupEntity(
+                Loc.GetString("energy-dome-no-power"),
+                generator,
+                user);
+            return false;
+        }
+
+        Toggle(generator);
+        return true;
+    }
+
+    public void Toggle(Entity<EnergyDomeGeneratorComponent> generator)
+    {
+        if (generator.Comp.Enabled)
+        {
+            TurnOff(generator, false);
+        }
+        else
+        {
+            TurnOn(generator);
+        }
+    }
+
+    public void TurnOn(Entity<EnergyDomeGeneratorComponent> generator)
+    {
+        if (generator.Comp.Enabled)
+            return;
+
+        generator.Comp.Enabled = true;
+        var newDome = Spawn(generator.Comp.DomePrototype, Transform(generator).Coordinates);
+
+        if (TryComp<EnergyDomeComponent>(newDome, out var domeComp))
+        {
+            domeComp.Generator = generator;
+        }
+
+        generator.Comp.SpawnedDome = newDome;
+        _audio.PlayPvs(generator.Comp.TurnOnSound, generator);
+    }
+
+    public void TurnOff(Entity<EnergyDomeGeneratorComponent> generator, bool startReloading)
+    {
+        if (!generator.Comp.Enabled)
+            return;
+
+        generator.Comp.Enabled = false;
+        QueueDel(generator.Comp.SpawnedDome);
+
+        _audio.PlayPvs(generator.Comp.TurnOffSound, generator);
+        if (startReloading)
+        {
+            generator.Comp.NextActivation = _gameTiming.CurTime + TimeSpan.FromSeconds(generator.Comp.ReloadSecond);
+            _audio.PlayPvs(generator.Comp.EnergyOutSound, generator);
+        }
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<EnergyDomeGeneratorComponent>();
+        while (query.MoveNext(out var uid, out var generator))
+        {
+            if (!generator.Enabled)
+                continue;
+
+            if (generator.SpawnedDome == null)
+                continue;
+
+            if (!_powerCell.TryUseCharge(uid, generator.Wattage * frameTime))
+            {
+                TurnOff(new Entity<EnergyDomeGeneratorComponent>(uid, generator), true);
+                continue;
+            };
+
+            _transform.SetCoordinates(generator.SpawnedDome.Value, Transform(uid).Coordinates);
```

Comment:
This is an ugly hack, please don't set coordinates every frame.

Reply:
Yeah, I know. I'll experiment more with parenting, but I need to somehow solve the problem with items "disappearing" from the world when they end up in containers